### PR TITLE
Ocrvs 2405 Correct record gateway, workflow, FHIR

### PR DESCRIPTION
**Pull request by tahmidrahman-dsi**
Correct record mutation example, correction information is sent in `correction` property inside `registration`  

```
mutation requestBirthRegistrationCorrection {
  requestBirthRegistrationCorrection(
    id: "bd91beb9-e951-437c-90a4-33a9cd2d9dd0"
    details: {
  createdAt: "2022-01-26T10:59:00.429Z",
  registration: {
    registrationNumber: "2022BYR7MLI",
    contact: MOTHER,
    contactPhoneNumber: "+260965656868",
    contactRelationship: "",
    _fhirID: "15d78710-8809-4a23-81db-45f917c15ac4",
    trackingId: "BYR7MLI",
    status: [
      {
        timestamp: "2022-01-26T10:59:00.431Z",
        timeLoggedMS: 6652,
      },
    ],
    attachments: [],
    correction: {
      location: {
        _fhirID: "63ee3076-4568-4cce-aa94-ad904b8ebfc8",
      },
      hasShowedVerifiedDocument: true,
      attestedAndCopied: true,
      requester: "MOTHER",
      reason: "CLERICAL ERROR",
      note: "Spelling mistakes",
      values: [
        {
          section: "child",
          fieldName: "name",
          oldValue: "Khaby Lame",
          newValue: "Khaby Lame Corrected"
        }, 
        {
          section: "mother",
          fieldName: "name",
          oldValue: "First Name Last Name",
          newValue: "Mother Family Name"
        }
      ]
    },
    draftId: "c41a95b6-4439-4fc7-9d56-6d3a8ea0f80b",
  },
  presentAtBirthRegistration: MOTHER,
  child: {
    name: [
      {
        use: "en",
        firstNames: "Khaby",
        familyName: "Lame Corrected",
      },
    ],
    gender: "male",
    birthDate: "2020-09-19",
    multipleBirth: 3,
    _fhirID: "cc57740e-8d6c-4147-9470-60f329f6854d",
  },
  attendantAtBirth: PHYSICIAN,
  birthType: SINGLE,
  weightAtBirth: 1.6,
  eventLocation: {
    _fhirID: "63ee3076-4568-4cce-aa94-ad904b8ebfc8",
  },
  mother: {
    nationality: ["ZMB"],
    identifier: [
      {
        id: "123456789",
        type: NATIONAL_ID,
      },
    ],
    name: [
      {
        use: "en",
        firstNames: "Mother",
        familyName: "Family Name",
      },
    ],
    maritalStatus: MARRIED,
    address: [
      {
        type: PLACE_OF_HERITAGE,
        line: ["", "", "", "", "", ""],
        country: "ZMB",
        state: "ec34cfe2-b566-4140-af22-71ff17d832d6",
        district: "1b49e1f7-7ca2-4417-8649-88b810289108",
      },
      {
        type: PERMANENT,
        line: ["", "", "", "", "", "", "URBAN"],
        country: "ZMB",
        state: "ec34cfe2-b566-4140-af22-71ff17d832d6",
        district: "1b49e1f7-7ca2-4417-8649-88b810289108",
      },
      {
        type: CURRENT,
        line: ["", "", "", "", "", "", "URBAN"],
        country: "ZMB",
        state: "ec34cfe2-b566-4140-af22-71ff17d832d6",
        district: "1b49e1f7-7ca2-4417-8649-88b810289108",
      },
    ],
    _fhirID: "0e50f510-aa22-47f7-9f72-90b51057e336",
  },
  _fhirIDMap: {
    composition: "c41a95b6-4439-4fc7-9d56-6d3a8ea0f80b",
    encounter: "e96e7b0f-53a4-4171-8021-ccb83d229390",
    eventLocation: "63ee3076-4568-4cce-aa94-ad904b8ebfc8",
    observation: {
      birthType: "2ec4507a-8f60-4f3b-ab1d-201bd716fdd6",
      weightAtBirth: "9c4b623d-923f-47ec-8647-8dbe6915e9c3",
      attendantAtBirth: "e046da21-6e9a-48b9-93ab-df09057f1f73",
      presentAtBirthRegistration: "4d07eb85-cd4e-446f-86fe-f0128729465e",
    },
  },
}
  )
}
```
After calling hearth, this causes mutating the existing resources, (i.e. in this example, it updates the names of child and mother Patient resources. Also keeps the record that which values are changed and also from which values are changed in the Task resource in input (holds the values from where those are changed) and output (holds the values that are changed) arrays. This mutation updates the Task like the following. 

```
{
    "_id" : ObjectId("61f12948f81189001c90a745"),
    "resourceType" : "Task",
    "status" : "requested",
    "code" : {
        "coding" : [ 
            {
                "system" : "http://opencrvs.org/specs/types",
                "code" : "BIRTH"
            }
        ]
    },
    "focus" : {
        "reference" : "Composition/c41a95b6-4439-4fc7-9d56-6d3a8ea0f80b"
    },
    "id" : "15d78710-8809-4a23-81db-45f917c15ac4",
    "identifier" : [ 
        {
            "system" : "http://opencrvs.org/specs/id/draft-id",
            "value" : "c41a95b6-4439-4fc7-9d56-6d3a8ea0f80b"
        }, 
        {
            "system" : "http://opencrvs.org/specs/id/birth-tracking-id",
            "value" : "BYR7MLI"
        }, 
        {
            "system" : "http://opencrvs.org/specs/id/birth-registration-number",
            "value" : "2022BYR7MLI"
        }
    ],
    "extension" : [ 
        {
            "url" : "http://opencrvs.org/specs/extension/contact-person",
            "valueString" : "MOTHER"
        }, 
        {
            "url" : "http://opencrvs.org/specs/extension/contact-relationship",
            "valueString" : ""
        }, 
        {
            "url" : "http://opencrvs.org/specs/extension/contact-person-phone-number",
            "valueString" : "+260965656868"
        }, 
        {
            "url" : "http://opencrvs.org/specs/extension/timeLoggedMS",
            "valueInteger" : 6652
        }, 
        {
            "url" : "http://opencrvs.org/specs/extension/requestingIndividual",
            "valueString" : "MOTHER"
        }, 
        {
            "url" : "http://opencrvs.org/specs/extension/regLastLocation",
            "valueReference" : {
                "reference" : "Location/976c4871-c4d1-46f7-b50b-09f346e05d55"
            }
        }, 
        {
            "url" : "http://opencrvs.org/specs/extension/regLastOffice",
            "valueReference" : {
                "reference" : "Location/fe68d322-4d98-4cdf-a320-2db9653f781d"
            }
        }, 
        {
            "url" : "http://opencrvs.org/specs/extension/regLastUser",
            "valueReference" : {
                "reference" : "Practitioner/7ce43df4-0692-424d-98ce-4155c2f18757"
            }
        }
    ],
    "lastModified" : "2022-01-26T13:27:08.355Z",
    "input" : [ 
        {
            "valueCode" : "child",
            "valueId" : "name",
            "type" : {
                "coding" : [ 
                    {
                        "system" : "http://terminology.hl7.org/CodeSystem/action-type",
                        "code" : "update"
                    }
                ]
            },
            "valueString" : "Khaby Lame"
        }, 
        {
            "valueCode" : "mother",
            "valueId" : "name",
            "type" : {
                "coding" : [ 
                    {
                        "system" : "http://terminology.hl7.org/CodeSystem/action-type",
                        "code" : "update"
                    }
                ]
            },
            "valueString" : "First Name Last Name"
        }
    ],
    "output" : [ 
        {
            "valueCode" : "child",
            "valueId" : "name",
            "type" : {
                "coding" : [ 
                    {
                        "system" : "http://terminology.hl7.org/CodeSystem/action-type",
                        "code" : "update"
                    }
                ]
            },
            "valueString" : "Khaby Lame Corrected"
        }, 
        {
            "valueCode" : "mother",
            "valueId" : "name",
            "type" : {
                "coding" : [ 
                    {
                        "system" : "http://terminology.hl7.org/CodeSystem/action-type",
                        "code" : "update"
                    }
                ]
            },
            "valueString" : "Mother Family Name"
        }
    ],
    "reason" : {
        "text" : "CLERICAL ERROR"
    },
    "note" : [ 
        {
            "text" : "Spelling mistakes",
            "time" : "Wed, 26 Jan 2022 13:27:08 GMT",
            "authorString" : ""
        }
    ],
    "businessStatus" : {
        "coding" : [ 
            {
                "system" : "http://opencrvs.org/specs/reg-status",
                "code" : "REQUESTED_CORRECTION"
            }
        ]
    },
    "meta" : {
        "lastUpdated" : "2022-01-26T13:27:08.367+00:00",
        "versionId" : "e23d17bc-d7fb-40b5-85d6-7bbe73b0e2e4"
    },
    "_transforms" : {
        "meta" : {
            "lastUpdated" : ISODate("2022-01-26T13:27:08.367Z")
        }
    },
    "_request" : {
        "method" : "PUT"
    }
}
```

To keep the information of "who is requesting the correction" is stored in an extension having an URL  "http://opencrvs.org/specs/extension/requestingIndividual". Reason is saved in `reason` property. Currently, the event is handled in workflow and also forwarded to openHIM.

NOTE: Merging this PR also require merging https://github.com/opencrvs/opencrvs-farajaland/pull/144

### CODE REVIEW

File: packages/workflow/src/features/events/handler.ts [hunk 1]
```diff
@@ -83,7 +87,14 @@ function detectEvent(request: Hapi.Request): Events {
                 return Events.BIRTH_WAITING_VALIDATION
               }
             } else {
-              return Events.BIRTH_MARK_CERT
+              if (
+                hasRegisterScope(request) &&
+                hasCorrectionEncounterSection(firstEntry as fhir.Composition)
```
Comment (tahmidrahman-dsi) [inline packages/workflow/src/features/events/handler.ts:92]: The payload should have a registration number, as the records are to be corrected can be found in ready to print tab

**Comment by Zangetsu101** [comment]
Should we send a notification for `REQUESTED_CORRECTION`? https://github.com/opencrvs/opencrvs-core/blob/04ae0f82c4e5754c71bea08f021d949cd743e648/packages/workflow/src/features/registration/utils.ts#L59

**Comment by tahmidrahman-dsi** [comment]
> Should we send a notification for `REQUESTED_CORRECTION`?
> 
> https://github.com/opencrvs/opencrvs-core/blob/04ae0f82c4e5754c71bea08f021d949cd743e648/packages/workflow/src/features/registration/utils.ts#L59

Thanks, @Zangetsu101 for pointing that out. I am not sure as it was not mentioned on the ticket. Let's ask @euanmillar about this

**Comment by euanmillar** [comment]
@tahmidrahman-dsi no notification is currently required

**Review by euanmillar — APPROVED** [review]
Great work Tahmid!!

**Pull request merged**